### PR TITLE
multipath-tools 0.9.8

### LINKED ISSUE #76: Socket activation doesn't play well with udev rule
Hi,

I'm working on [implementing real socket activation](https://salsa.debian.org/linux-blocks-team/multipath-tools/-/merge_requests/11) for `multipathd.service` on Debian/Ubuntu, and I'm having problems with how the current `60-multipath.rules` causes the socket to trigger the service start even when there's no multipath device in the system.

The problem seems to happen because the udev rules will invoke `multipath`, like in:

https://github.com/opensvc/multipath-tools/blob/f3004b45e7f8266a3f00b146d4742821d04b7940/multipath/multipath.rules.in#L32-L35

This will cause `multipathd.socket` to be notified, which finally makes the service start.  As I said, it doesn't matter if there is a multipath device in the system: the service will always start.

I'm thinking about possible ways to overcome this limitation, but it would be extremely helpful to hear your opinions as well.

Thanks in advance.

### LINKED ISSUE #80: 0.9.7: test suite build fails with gcc 14.x
Looks like last version build fails with latest gcc 14.x which is now used in fedora rawhide.
<details>
<summary>Build fails with</summary>

```console
make[1]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/multipath-tools-0.9.7/multipathd'
make[1]: Entering directory '/home/tkloczko/rpmbuild/BUILD/multipath-tools-0.9.7/tests'
building alias.o because of alias.c ../libmpathutil/strbuf.h ../libmpathutil/util.h ../libmultipath/alias.h test-log.h globals.c ../libmultipath/defaults.h ../libmultipath/structs.h ../libmultipath/prio.h ../libmultipath/checkers.h ../libmultipath/list.h ../libmpathutil/vector.h ../libmultipath/byteorder.h ../libmultipath/generic.h ../libmultipath/config.h ../libmpathutil/globals.h ../libmpathutil/debug.h ../libmpathutil/log_pthread.h ../libmultipath/alias.c ../libmpathutil/uxsock.h ../libmultipath/file.h ../libmultipath/devmapper.h ../libmultipath/autoconfig.h ../libmpathutil/time-util.h ../libmultipath/lock.h
alias.c:1766:12: error: function declaration isn’t a prototype [-Werror=strict-prototypes]
 1766 | static int test_get_user_friendly_alias()
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
alias.c:1984:12: error: function declaration isn’t a prototype [-Werror=strict-prototypes]
 1984 | static int test_bindings_order()
      |            ^~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[1]: *** [Makefile:74: alias.o] Error 1
make[1]: Target 'all' not remade because of errors.
rm alias.o.wrap
make[1]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/multipath-tools-0.9.7/tests'
make: *** [Makefile:121: test] Error 2
```
</details>

### PR DESCRIPTION
**Pull request by mwilck**
Hi Christophe,

here's a PR for 0.9.8. As usual, all patches have been reviewed, except the ones marked with (*) below, which are some GitHub action changes and 2 minor documentation fixes.

(**Side note:** we should start thinking about what the next version after 0.9.9 should be).

## News

* Socket activation via `multipathd.socket` has been disabled by default because it has undesirable side effects (fixes #76, at least partially).
* The `restorequeueing` CLI command now only enables queueing if `disablequeueing` had been sent before.
* multipathd now tracks the queueing mode of maps in its internal features string. This is helpful to ensure that maps have the desired queuing status. Without this, it can happen that a map remains in queueing state even after the no_path_retry timeout has expired.
* multipathd's map flushing code has been reworked to avoid hangs if there are no paths but outstanding IO. This allows "multipath -F" to retry a flush using the daemon rather than retrying locally.
* Error messages sent from multipathd to the command line client have been improved. The user will now see messages like "map or partition in use" or "device not found" instead of just "fail".
* Enabled `-D_FILE_OFFSET_BITS=64` to fix issues with emulated 32-bit environments in the GitHub CI, so that we can now run our CI in arm/v7.

## Bug fixes

* A segfault in the 0.9.7 autoresize code has been fixed.
* Fixed a bug introduced in 0.9.6 that had caused map reloads being ommitted when path priorities changed.
* Fixed compilation with gcc 14. (Fixes #80)
* Minor fixes for issues detected by coverity.
* Spelling fixes and other minor fixes.

## CI

* Added the [check-spelling](https://github.com/marketplace/actions/check-spelling) GitHub action to avoid spelling errors in user-facing files in the future.
* Various improvements and updates for GitHub actions 

## Patch breakdown

@bmarzins (26):
      multipathd: Make sure to disable queueing if recovery has failed.
      multipathd: don't modify the multipath device on show commands
      libmultipath: keep track of queueing state in features
      multipathd: only restore queueing if it has been disabled first
      multipathd: remove nopath flushing code from flush_map()
      multipathd: make flush_map() delete maps like the multipath command
      multipathd: disable queueing when removing unknown maps
      multipathd: simplify cli_del_map()
      libmultipath: make _dm_flush_map() return an enum
      libmultipath: make dm_remove_partmaps() a static function
      multipathd: always start failure replies with "fail\n"
      multipathd: print extra default reply msg for busy devices
      multipathd: handle busy devices in cli_del_maps()
      libmultipath: print error when find_mp_by_str() fails.
      multipathd: don't open code find_mp_by_str() in client handers
      multipathd: make cli_handlers check for paths by devt and dev
      multipathd: add cli_handler reply message for missing devs
      libdmmp: handle failures in _process_cmd
      multipath: get rid of unnecessary retries variable
      multipath: Don't locally retry deletgated remove failures
      multipath: if delegation times out mark as not delegated
      multipathd: sync features on flush_map failure corner case
      multipathd: fix null pointer dereference in uev_update_path
      multipathd: fix auto-resize configuration
      libmultipath: fix displaying auto_resize config setting
      multipath-tools tests: add void parameter to functions

@brianatpurestorage (2):
      multipathd: the local path change is not considered
      libmultipath: stop PURE FlashArray from detecting priority

@jsoref (9):
      spelling: anymore
      spelling: cannot
      spelling: case-insensitive
      spelling: case-sensitive
      spelling: configuration
      spelling: correctly
      spelling: numerically
      spelling: preexisting
      spelling: than

@mwilck (29):
     libmultipath: avoid temporarily enabling queueing on reload
      libmultipath: fix ANA prioritizer enablement logic
      multipathd: init_unwinder: protect pthread_cond_wait() call with a loop
      libmultipath: is_uevent_busy(): check servicing_uev under lock
      libmultipath: tur: protect pthread_cond_timedwait with a loop
      multipathd: make update_prio static, and rename refresh_all param
      multipathd: don't activate socket activation by default
      multipath: udev rules: use configured $(bindir) in udev rules
      multipath-tools: Makefile.inc: set _FILE_OFFSET_BITS=64
      GitHub workflows: update workflow events (*)
      GitHub workflows: update distros to run for (*)
      GitHub workflows: split multiarch workflow in rolling/stable (*)
      GitHub Workflows: run expensive workflows only on relevant changes (*)
      GitHub workflows: add spell checker (*)
      Spelling fixes found by check-spelling action (*)
      GitHub Workflows: abi: only run if C code has changed (*)
      GitHub workflows: abi: error if reference ABI can't be downloaded (*)
      GitHub workflows: run on PRs against "queue", too (*)
      multipath.conf.5: fix documentation for find_multipaths (fixes #75) (*)
      libmultipath: hwtable: fix fast_io_fail for Infinibox
      11-dm-mpath.rules: fix list of imported properties
      11-dm-mpath.rules: use import logic like 13-dm-disk.rules
      11-dm-mpath.rules: don't import DM_UDEV_DISABLE_OTHER_RULES_FLAG
      11-dm-mpath.rules: handle reloads during coldplug events
      11-dm-mpath.rules: don't save DM_UDEV_DISABLE_OTHER_RULES_FLAG_OLD
      11-dm-mpath.rules: clear DM_DISABLE_OTHER_RULES_FLAG for coldplug events
      libmultipath: bump version to 0.9.8 (*)
      multipath-tools: added NEWS.md (*)

@xosevp (3):
      multipath-tools: update ml
      multipath-tools: fix an assignment ambiguity
      multipath-tools: remove extra hyphen from CFLAGS std option

### CODE REVIEW

File: .github/actions/spelling/expect.txt [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+reconfig
+redhat
+restorequeueing
+retrigger
+rhabarber
+rootprefix
+rootprefixdir
+rport
+rtpi
+SANtricity
```
Comment (jsoref) [inline .github/actions/spelling/expect.txt:166]: Similarly, this is parsed as `Ntricity` -- the `SA` is split off and then ignored because it's too short.

**Comment by jsoref** [comment]
>. multipathd now tracks the queueing mode of maps in ~it's~ _its_ internal features string. This is ~helful~ _helpful_ to ensure that maps have the desired queuing status. Without this, it can happen that a map remains in _the_ queueing state even after the no_path_retry timeout has expired.

File: .github/workflows/spelling.yml [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      # note: If you use only_check_changed_files, you do not want cancel-in-progress
+      cancel-in-progress: true
+    steps:
+    - name: check-spelling
+      id: spelling
+      uses: check-spelling/check-spelling@main
+      with:
+        suppress_push_for_open_pull_request: ${{ github.actor != 'dependabot[bot]' && 1 }}
+        checkout: true
+        check_file_names: 0
```
[the remark this answers is not in the document]
Comment (mwilck) [reply .github/workflows/spelling.yml:111]: I put it here because check-spelling showed errors related to file name checking. I didn't investigate further.

File: .github/workflows/spelling.yml [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        spell_check_this: check-spelling/spell-check-this@prerelease
```
[the remark this answers is not in the document]
Comment (mwilck) [reply .github/workflows/spelling.yml]: ok

File: .github/workflows/spelling.yml [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        post_comment: 0
+        use_magic_file: 1
+        report-timing: 1
```
[the remark this answers is not in the document]
Comment (mwilck) [reply .github/workflows/spelling.yml]: ok

File: .github/actions/spelling/expect.txt [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+ata
+autoconfig
+autodetected
+barbie
+BINDIR
+blkid
+bmarzins
+cciss
+CFLAGS
+CLARiiON
```
[the remark this answers is not in the document]
Comment (mwilck) [reply .github/actions/spelling/expect.txt]: fixeed

**Comment by mwilck** [comment]
> > . multipathd now tracks the queueing mode of maps in ~it's~ _its_ internal features string. This is ~helful~ _helpful_ to ensure that maps have the desired queuing status. Without this, it can happen that a map remains in _the_ queueing state even after the no_path_retry timeout has expired.

Sic :grin: You don't need no spell-checker, do you?
Thanks for your review!

File: .github/workflows/spelling.yml [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    - name: check-spelling
+      id: spelling
+      uses: check-spelling/check-spelling@main
+      with:
+        suppress_push_for_open_pull_request: ${{ github.actor != 'dependabot[bot]' && 1 }}
+        checkout: true
+        spell_check_this: opensvc/multipath-tools@master
+        post_comment: 0
+        use_magic_file: 1
+        report-timing: 0
```
Comment (jsoref) [inline .github/workflows/spelling.yml:114]: Similarly, you can just omit this line (the default is not to do it).

This is unlike `post_comment` where the default _was_ to do it, but it turned out that to behave well it had to be split to a separate pass.

Comment (mwilck) [reply .github/workflows/spelling.yml:114]: Unless it does harm, I'd like to keep it as-is for now.

Comment (jsoref) [reply .github/workflows/spelling.yml:114]: It's harmless.

**Comment by mwilck** [comment]
Pushed a minor change to the `_FILE_OFFSET_BITS=64` patch: use `$(OPTFLAGS)` in the cpp command (although it's rather pointless) to avoid warning about missing optimization flags when building the unit tests.

File: .github/workflows/spelling.yml (hunk 6 above)
Comment (mwilck) [reply .github/workflows/spelling.yml:114]: I guess leaving this in won't do harm either.

**Comment by mwilck** [comment]
Pushed another patch after review from @bmarzins.

Once [[PATCH v3] 11-dm-mpath.rules: handle reloads during coldplug events](https://lore.kernel.org/dm-devel/20240210171327.11811-1-mwilck@suse.com/T/#ebbb39f0f020fe88175bae5ea3882b87b5498ed6f) gets reviewed, I can add the udev rule series, too.

**Comment by mwilck** [comment]
The CI run for "multiarch test for rolling distros / build-current (debian-sid, arm/v7)" has apparently failed because of an issue with downloading the test container. @cvaroqui please restart [the workflow](https://github.com/opensvc/multipath-tools/actions/runs/7898332153/job/21555531411?pr=81) if you have time.

**Review by bmarzins — APPROVED** [review]
Looks good to me.

**Comment by mwilck** [comment]
Pushed the the patch series for 11-dm-mpath.rules now, as the series has been completely reviewed.

**Comment by mwilck** [comment]
Pushed version bump to 0.9.8

**Comment by mwilck** [comment]
@cvaroqui, I believe this should be good for merging now.

**Comment by mwilck** [comment]
Again, the workflow failures seem to have been caused by problems pulling the test containers.
